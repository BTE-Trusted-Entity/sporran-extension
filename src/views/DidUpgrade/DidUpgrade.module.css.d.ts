export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const details: string;
export const detail: string;
export const detailName: string;
export const detailValue: string;
export const info: string;
export const paypal: string;
export const buttonsLine: string;
export const cancel: string;
export const submit: string;
export const footnote: string;
