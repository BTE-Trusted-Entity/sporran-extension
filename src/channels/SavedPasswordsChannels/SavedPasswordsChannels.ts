import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
