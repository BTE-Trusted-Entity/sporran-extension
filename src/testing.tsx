export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
