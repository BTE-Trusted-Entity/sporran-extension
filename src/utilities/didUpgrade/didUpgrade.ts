import BN from 'bn.js';
import {
  Blockchain,
  ConfigService,
  Did,
  DidDocument,
  DidUri,
  KiltAddress,
  KiltKeyringPair,
  SignExtrinsicCallback,
  SubmittableExtrinsic,
} from '@kiltprotocol/sdk-js';

import { useMemo } from 'react';

import {
  getIdentityCryptoFromSeed,
  getLightDidFromSeed,
} from '../identities/identities';
import { parseDidUri } from '../did/did';
import { useAsyncValue } from '../useAsyncValue/useAsyncValue';
import { useAddressBalance } from '../../components/Balance/Balance';
import { getDepositDid } from '../getDeposit/getDeposit';
import { makeFakeIdentityCrypto } from '../makeFakeIdentityCrypto/makeFakeIdentityCrypto';

interface DidTransaction {
  extrinsic: SubmittableExtrinsic;
  did: DidUri;
}

export async function getTransaction(
  lightDidDocument: DidDocument,
  keypair: KiltKeyringPair,
  sign: SignExtrinsicCallback,
  submitter?: KiltAddress,
): Promise<DidTransaction> {
  const { fullDid: did } = parseDidUri(lightDidDocument.uri);

  const extrinsic = await Did.getStoreTx(
    lightDidDocument,
    submitter || keypair.address,
    async (input) => sign({ ...input, did }),
  );

  return { extrinsic, did };
}

export async function getFee(): Promise<BN> {
  const { keypair, sign, fakeSeed } = makeFakeIdentityCrypto();
  const document = getLightDidFromSeed(fakeSeed);

  const api = ConfigService.get('api');

  const { extrinsic } = await getTransaction(document, keypair, sign);
  const signed = await extrinsic.signAsync(keypair);
  const extrinsicFee = (await signed.paymentInfo(keypair)).partialFee;

  const didCreationFee = api.consts.did.fee;

  return extrinsicFee.add(didCreationFee);
}

export function useKiltCosts(
  address: string,
  did: DidUri | undefined,
): {
  fee?: BN;
  deposit?: BN;
  total?: BN;
  insufficientKilt: boolean;
} {
  const fee = useAsyncValue(getFee, []);
  const deposit = useAsyncValue(getDepositDid, [did]);

  const total = useMemo(
    () => (fee && deposit ? fee.add(deposit.amount) : undefined),
    [deposit, fee],
  );

  const balance = useAddressBalance(address);
  const insufficientKilt = Boolean(
    total && balance && balance.transferable.lt(total),
  );

  return { fee, deposit: deposit?.amount, total, insufficientKilt };
}

const currentTx: Record<string, DidTransaction> = {};

export async function sign(seed: Uint8Array): Promise<string> {
  const document = getLightDidFromSeed(seed);
  const { keypair, sign } = await getIdentityCryptoFromSeed(seed);
  const { extrinsic, did } = await getTransaction(document, keypair, sign);

  const signed = await extrinsic.signAsync(keypair);

  const hash = signed.hash.toHex();
  currentTx[hash] = { extrinsic: signed, did };
  return hash;
}

export async function submit(hash: string): Promise<DidUri> {
  const { extrinsic, did } = currentTx[hash];
  await Blockchain.submitSignedTx(extrinsic);
  delete currentTx[hash];

  return did;
}
