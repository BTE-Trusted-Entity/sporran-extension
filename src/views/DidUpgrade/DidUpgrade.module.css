.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.costs {
  composes: smallText from '../../components/Typography/Typography.module.css';
  font-size: 1.25rem;
  margin: 0.25rem 0 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.details {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.5rem 0 1.5rem;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin: 1.5rem 0 0;
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
