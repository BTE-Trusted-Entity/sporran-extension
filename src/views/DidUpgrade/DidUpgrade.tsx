import { FormEvent, useCallback, useRef, useState } from 'react';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './DidUpgrade.module.css';

import { Identity } from '../../utilities/identities/types';
import {
  PasswordError,
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import {
  sign,
  submit,
  useKiltCosts,
} from '../../utilities/didUpgrade/didUpgrade';
import {
  getIdentityDid,
  saveIdentity,
} from '../../utilities/identities/identities';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { KiltCurrency } from '../../components/KiltCurrency/KiltCurrency';
import {
  asKiltCoins,
  KiltAmount,
} from '../../components/KiltAmount/KiltAmount';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { ExplainerModal } from '../../components/ExplainerModal/ExplainerModal';
import { Stats } from '../../components/Stats/Stats';
import { paths } from '../paths';

interface Props {
  identity: Identity;
}

export function DidUpgrade({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { address } = identity;
  const did = getIdentityDid(identity);

  const { fee, deposit, total, insufficientKilt } = useKiltCosts(address, did);
  const [txHash, setTxHash] = useState<string>();

  const [submitting, setSubmitting] = useState(false);
  const [status, setStatus] = useState<'pending' | 'success' | 'error' | null>(
    null,
  );

  const passwordField = usePasswordField();

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      try {
        const { seed } = await passwordField.get(event);

        setSubmitting(true);
        setStatus('pending');

        const hash = await sign(seed);
        setTxHash(hash);

        const did = await submit(hash);
        await saveIdentity({ ...identity, did });

        setStatus('success');
      } catch (error) {
        if (error instanceof PasswordError) {
          return;
        }
        setSubmitting(false);
        setStatus('error');
      }
    },
    [identity, passwordField],
  );

  const closeModal = useCallback(() => {
    setStatus(null);
  }, []);

  const portalRef = useRef<HTMLDivElement>(null);

  if (!(fee && deposit && total)) {
    return null; // blockchain data pending
  }

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>{t('view_DidUpgrade_heading')}</h1>
      <p className={styles.subline}>{t('view_DidUpgrade_subline')}</p>

      <IdentitySlide identity={identity} />

      <p className={styles.costs}>
        {t('view_DidUpgrade_total')}
        <KiltAmount amount={total} type="costs" smallDecimals />
      </p>
      <p className={styles.details}>
        <ExplainerModal label={t('view_DidUpgrade_info')} portalRef={portalRef}>
          {t('view_DidUpgrade_deposit_explainer')}
        </ExplainerModal>
        {t('view_DidUpgrade_deposit')}
        {asKiltCoins(deposit, 'costs')} <KiltCurrency />
        {t('view_DidUpgrade_fee')}
        {asKiltCoins(fee, 'costs')} <KiltCurrency />
      </p>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button
          type="submit"
          className={styles.submit}
          disabled={submitting || insufficientKilt}
        >
          {t('common_action_sign')}
        </button>
        <output
          className={styles.errorTooltip}
          hidden={!insufficientKilt || Boolean(status)}
        >
          {t('view_DidUpgrade_insufficientFunds', [
            asKiltCoins(total, 'costs'),
          ])}
        </output>
      </p>

      {status && (
        <TxStatusModal
          identity={identity}
          status={status}
          txHash={txHash}
          onDismissError={closeModal}
          messages={{
            pending: t('view_DidUpgrade_pending'),
            success: t('view_DidUpgrade_success'),
            error: t('view_DidUpgrade_error'),
          }}
        />
      )}

      <div ref={portalRef} />

      <LinkBack />
      <Stats />
    </form>
  );
}
