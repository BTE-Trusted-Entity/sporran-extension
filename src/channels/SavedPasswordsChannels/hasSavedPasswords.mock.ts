import { act } from '@testing-library/react';

import { hasSavedPasswordsChannel } from './SavedPasswordsChannels';

jest.mock('./SavedPasswordsChannels');
const hasSavedPasswordsPromise = Promise.resolve(false);
(hasSavedPasswordsChannel.get as jest.Mock).mockReturnValue(
  hasSavedPasswordsPromise,
);

export async function waitForHasSavedPasswords(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}
