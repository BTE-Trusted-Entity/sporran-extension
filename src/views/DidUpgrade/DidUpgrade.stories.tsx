import { Meta } from '@storybook/react';
import { MemoryRouter, Route } from 'react-router-dom';

import { identitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';
import { paths } from '../paths';

import { DidUpgrade } from './DidUpgrade';

export default {
  title: 'Views/DidUpgrade',
  component: DidUpgrade,
} as Meta;

export function Template(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        '/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/did/upgrade/kilt',
      ]}
    >
      <Route path={paths.identity.did.upgrade.kilt}>
        <DidUpgrade
          identity={
            identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
          }
        />
      </Route>
    </MemoryRouter>
  );
}
