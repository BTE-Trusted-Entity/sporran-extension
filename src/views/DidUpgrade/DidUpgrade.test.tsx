import { BalanceUtils } from '@kiltprotocol/sdk-js';

import { identitiesMock as identities, render } from '../../testing/testing';

import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import { useKiltCosts } from '../../utilities/didUpgrade/didUpgrade';

import { DidUpgrade } from './DidUpgrade';

jest.mock('../../utilities/didUpgrade/didUpgrade', () => ({
  useKiltCosts: jest.fn(),
}));

jest.mocked(useKiltCosts).mockReturnValue({
  fee: BalanceUtils.toFemtoKilt(0.01),
  deposit: BalanceUtils.toFemtoKilt(1),
  total: BalanceUtils.toFemtoKilt(1.01),
  insufficientKilt: false,
});

describe('DidUpgrade', () => {
  it('should render', async () => {
    const { container } = render(
      <DidUpgrade
        identity={
          identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
