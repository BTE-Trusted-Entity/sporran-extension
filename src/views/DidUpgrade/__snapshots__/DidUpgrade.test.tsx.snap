// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidUpgrade should render 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Get your DID
    </h1>
    <p
      class="subline"
    >
      Sign
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Light DID Identity
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Total cost: 
      <span
        class="amount"
      >
        1
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
      <button
        aria-label="Information"
        class="show"
        type="button"
      />
       Deposit 
      1.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       + fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 1.0100 KILTs
      </output>
    </p>
    <div />
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
